### ISSUE TITLE
Unable to reproduce raw scores using SpliceAI python package

### ISSUE THREAD

**Issue by rickymagner** [broadinstitute/SpliceAI-lookup#110]
Hi, I was using the `spliceai` Python package to try to reproduce some of the raw scores for variants I'm looking at through the lookup web portal. In particular, [this](https://spliceailookup.broadinstitute.org/#variant=chr17-43090942-A-T&hg=38&bc=basic&distance=10000&mask=1&ra=0) variant (chr17-43090942-A-T) with max distance set to 10000 shows (correctly) there is a donor site for the BRCA1 transcript at 43090943 with probability 0.99 in the IGV window.

However, when trying to reproduce this value using the package using the example from the package README, I don't get any donor sites that have high enough probability to be considered canonical. In particular:
```
from keras.models import load_model
from pkg_resources import resource_filename
from spliceai.utils import one_hot_encode
import numpy as np
import pysam


fasta_path = <my local hg38.fa path>
with pysam.FastaFile(fasta_path) as fasta:
    test_ref = fasta.fetch('chr17', 43090942-1-10_000, 43090942+10_000)

ref_test_prob = np.mean([splice.models[m].predict(one_hot_encode(5000*'N' + test_ref + 5000*'N')[None, :]) for m in range(5)], axis=0)
ref_donor_prob = ref_test_prob[0, :, 2] 
print(sum(ref_donor_prob > .5))
```
outputs 0, i.e. there are no reference bases in this window with > 0.5 probability of being a donor. What's going on here? How are the raw scores being computed in the web portal?

**Comment by bw2** [comment]
Hi @rickymagner , can you try running this inside the docker.io/weisburd/spliceai-38:latest image which has the [SpliceAI version](https://github.com/bw2/spliceai), gene annotation file, and reference fasta used by the site's backend to call spliceai @  
https://github.com/broadinstitute/SpliceAI-lookup/blob/master/google_cloud_run_services/server.py#L362-L377
?

**Comment by rickymagner** [comment]
Thanks for getting back to me @bw2. I tried running your docker but it seems to hang:
```
> podman run -it --rm --platform linux/amd64 -v $(pwd):/test_spliceai weisburd/spliceai-38
 * Serving Flask app 'server'
 * Debug mode: on
```
Where can I find the server, and how should I run the code snippet above there? Also you linked to a block of code that looks like it gets delta scores, but I want to produce the raw SpliceAI scores that are depicted e.g. in the IGV output at the bottom of the web UI.

**Comment by bw2** [comment]
When you run it, it starts the API server by default. Try
```
podman run -it --rm --platform linux/amd64 -v $(pwd):/test_spliceai   weisburd/spliceai-38   /bin/bash
```
to get an interactive shell inside the container. 

That delta function should include the raw scores and not just delta scores (this was one of the main changes on my fork of SpliceAI).

**Comment by rickymagner** [comment]
Ok, it looks like I was able to run my own script and get similar scores matching the ones in the web browser, both for my input variant (which I realize I forgot to reverse complement for that gene), and another that wasn't working before (on the + strand). 

What do you recommend as far as being able to have the same functionality for writing my own scripts to use the raw scores from SpliceAI? Would it make sense to use some parts of your fork for this if you have an API I could use (that code block looked steeped in server config to be able to extract easily)? Or do you recommend I use the official ILMN package but in a different way than I attempted above?

**Comment by rickymagner** [comment]
Also for my use case, I'd like to be able to batch a couple of variants at once hopefully to get a speedup. I tried using a `np.stack` to combine my ref + alt sequences into a tensor of shape (2, ..., 4) after one-hot encoding, but it seemed to give an error. Is it possible to do something to this effect and is it worth it for performance?

**Comment by bw2** [comment]
Glad it's working. I would recommend either running your batched analysis inside the docker container or copying the annotation file(s) out of the docker image to your local disk, cloning my fork, and then writing your own python script that batch-processes your variants through  

 ```
# generate error message if variant falls outside annotated exons or introns
    record = VariantRecord(chrom, pos, ref, alt)
    try:
        scores = get_delta_scores(
            record,
            SPLICEAI_ANNOTATOR[(genome_version, basic_or_comprehensive_param)],
            distance_param,
            mask_param)
    except Exception as e:
        print(f"ERROR while computing SpliceAI scores for {variant}: {e}")
        traceback.print_exc()
        return {
            "variant": variant,
            "source": "spliceai",
            "error": f"{type(e)}: {e}",
        }
```

I don't know about `np.stack` and whether it can improve performance.

**Comment by rickymagner** [comment]
Thanks for the suggestion. A few more questions:
- What would be the annotations files required if I wanted to copy to another environment where I can't use the docker image?
- There are parameters for `get_delta_scores` related to the setup of the server. How should I set those, e.g. `SPLICEAI_ANNOTATOR` etc to get a minimal amount of code that can run `get_delta_scores` on its own?
- Do you know why your branch behaves differently than Illumina's official package when determining these raw scores? What did you change to get the raw scores to look more reasonable than theirs?

**Comment by bw2** [comment]
See how `SPLICEAI_ANNOTATOR` is defined to understand the files and where they're located in the docker image. This is unrelated to web server functionality - it's part of the core inputs to SpliceAI. 
To understand what was changed in https://github.com/bw2/spliceai I would recommend looking at the github commit history and/or doing a diff with the matching files in the original Illumina version.
